volume/service/db.go

```diff
@@ -74,11 +81,34 @@ func removeMeta(tx *bolt.Tx, name string) error {
 // listMeta is used during restore to get the list of volume metadata
 // from the on-disk database.
 // Any errors that occur are only logged.
-func listMeta(tx *bolt.Tx) []volumeMetadata {
+func listMeta(tx *bolt.Tx, rootPath string) []volumeMetadata {
 	var ls []volumeMetadata
 	b := tx.Bucket(volumeBucketName)
 	b.ForEach(func(k, v []byte) error {
 		if len(v) == 0 {
+			//Given that emptying the metadata.db of a volume does not affect the use of the volume,
+			//back up the database to facilitate restart the docker daemon.
+			defer func() {
+				if v := recover(); v != nil {
+					// Format the date and time, prevent overwriting
+					now := time.Now()
+					dateString := now.Format("20060102150405")
+					dbPath := filepath.Join(rootPath, "volumes/metadata.db")
+					errPath := filepath.Join(rootPath, "volumes/metadata.db.err."+dateString)
+					if err := os.Rename(dbPath, errPath); err != nil {
+						logrus.Errorf("os.Rename err: %v", err)
+					}
+					logrus.Errorf("Error while reading volume metadata Bucket is nil, The metadata.db backup: %v", errPath)
+					logrus.Errorf("List stack:\n %s \n", string(debug.Stack()))
+					fmt.Printf("Error while reading volume metadata Bucket is nil, The metadata.db backup: %v", errPath)
+					os.Exit(1)
```

Comment:
Perhaps you pushed while still working on this, but we moved away from direct imports  of logrus where possible, and instead you can use the `log` package;  e.g.


```go
log.L.Errorf(...)
```

or 

```go
log.G(context.TODO()).Errorf(.....
```

Reply:
Thanks a lot, I modified it